pkg/tabulator/tabstate.go

```diff
@@ -183,80 +197,133 @@ func Update(ctx context.Context, client gcs.ConditionalClient, mets *Metrics, co
 	}(ctx)
 
 	// Set up threads
-	var active stringset.Set
-	var waiting stringset.Set
-	var lock sync.Mutex
+	var activeGroup stringset.Set
+	var activeGroupLock sync.Mutex
+	groups := make(chan *configpb.TestGroup)
+
+	tabs := make(chan writeTask)
+	var tabLock sync.Mutex
+
+	read := func(ctx context.Context, log *logrus.Entry, group *configpb.TestGroup) error {
+		activeGroupLock.Lock()
+		ok := activeGroup.Add(group.Name)
+		activeGroupLock.Unlock()
+		if !ok {
+			q.Fix(group.Name, time.Now().Add(freq/10), false)
+			return nil
+		}
+
+		fromPath, err := updater.TestGroupPath(configPath, gridPathPrefix, group.Name)
+		if err != nil {
+			return fmt.Errorf("can't make tg path %q: %w", group.Name, err)
+		}
+
+		log.WithFields(logrus.Fields{
+			"from": fromPath.String(),
+		}).Info("Reading state")
 
-	channel := make(chan *configpb.TestGroup)
+		grid, _, err := gcs.DownloadGrid(ctx, client, *fromPath)
+		if err != nil {
+			return fmt.Errorf("downloadGrid(%s): %w", fromPath, err)
+		}
 
-	update := func(log *logrus.Entry, group *configpb.TestGroup) error {
+		tabLock.Lock()
+		// lock out all other readers so that all these tabs get handled as soon as possible
+		var wg sync.WaitGroup
+		wg.Add(len(groupToTabs[group.Name]))
 		for _, view := range groupToTabs[group.Name] {
-			log := log.WithField("dashboard", view.dashboard.Name).WithField("tab", view.tab.Name)
-			fromPath, err := updater.TestGroupPath(configPath, gridPathPrefix, view.tab.TestGroupName)
-			if err != nil {
-				return fmt.Errorf("can't make tg path %q: %w", view.tab.TestGroupName, err)
-			}
-			toPath, err := TabStatePath(configPath, tabsPathPrefix, view.dashboard.Name, view.tab.Name)
-			if err != nil {
-				return fmt.Errorf("can't make dashtab path %s/%s: %w", view.dashboard.Name, view.tab.Name, err)
-			}
-			log.WithFields(logrus.Fields{
-				"from": fromPath.String(),
-				"to":   toPath.String(),
-			}).Info("Calculating state")
-
-			err = createTabState(ctx, log, client, view.tab, group, *fromPath, *toPath, confirm, dropEmptyCols, calculateStats, useTabAlertSettings)
-			if err != nil {
-				return fmt.Errorf("can't calculate state: %w", err)
-			}
+			out := view
+			out.data = proto.Clone(grid).(*statepb.Grid)
```

Comment:
Seems a waste of inflate N times? Isn't this what we spend most of our CPU time doing?

Reply:
Not sure what you mean. Deep-copying the result is cheaper than inflating it each time.

Reply:
This code downloads the proto once and then inflates it N times.

Reply:
Ah, yes, of course it does. I have no idea why I didn't see that before. I'll look into passing inflated columns or something similar

Reply:
I can't reasonably fix this until "--filter-columns" is dropped as a feature flag. Will have to address this in a later PR.